File: packages/host-wire/src/capabilities.ts
import { fail } from "./errors.js";
import { capabilitiesArray, controlFree, inputObject } from "./guards.js";
export const DEVICE_CAPABILITIES = [
	"sessions.read",
	"sessions.prompt",
	"sessions.control",
	"sessions.manage",
	"bash.run",
	"term.open",
	"term.input",
	"term.resize",
	"files.read",
	"files.write",
	"files.list",
	"files.diff",
	"agents.control",
	"audit.read",
	"config.read",
	"catalog.read",
	"config.write",
	"broker.read",
	"usage.read",
	"preview.read",
	"preview.control",
	"preview.input",
	"ci.trigger",
] as const;
export type DeviceCapability = (typeof DEVICE_CAPABILITIES)[number];
export const PROTOCOL_FEATURES = [
	"resume",
	"host.watch",
	"session.watch",
	"session.state",
	"session.delta",
	"session.observer",
	"controller.lease",
	"prompt.lease",
	"prompt.images",
	"transcript.images",
	"transcript.search",
	"transcript.page",
	"project.reveal",
	"agent.lifecycle",
	"agent.progress",
	"agent.event",
	"agent.transcript",
	"terminal.io",
	"files.list",
	"files.search",
	"files.diff",
	"audit.tail",
	"catalog.metadata",
	"settings.metadata",
	"preview.control",
	"runtime.adapters",
	"workspace.lifecycle",
	"cluster.operator",
] as const;
export type ProtocolFeature = (typeof PROTOCOL_FEATURES)[number];
export const REMOTE_DEFAULT_CAPABILITIES = [
	"sessions.read",
	"sessions.prompt",
	"sessions.control",
	"sessions.manage",
	"agents.control",
] as const;
export interface Capabilities {
	client: string[];
	server?: string[];
}
export function decodeCapabilities(value: unknown, path = "capabilities"): Capabilities {
	const input = inputObject(value);
	const client = capabilitiesArray(input.client, `${path}.client`);
	const server = input.server === undefined ? undefined : capabilitiesArray(input.server, `${path}.server`);
	for (const [i, capability] of client.entries())
		if (!(DEVICE_CAPABILITIES as readonly string[]).includes(capability))
			fail("INVALID_FRAME", "unknown device capability", `${path}.client[${i}]`);
	if (server !== undefined)
		for (const [i, capability] of server.entries())
			if (!(DEVICE_CAPABILITIES as readonly string[]).includes(capability))
				fail("INVALID_FRAME", "unknown device capability", `${path}.server[${i}]`);
	return server === undefined ? { client } : { client, server };
}
export function decodeFeatureList(value: unknown, path: string): string[] {
	const result = capabilitiesArray(value, path);
	for (const [i, feature] of result.entries()) controlFree(feature, `${path}[${i}]`, 128);
	return result;
}
export function decodeNegotiatedFeatureList(value: unknown, path: string): ProtocolFeature[] {
	const result = decodeFeatureList(value, path);
	for (const [i, feature] of result.entries())
		if (!(PROTOCOL_FEATURES as readonly string[]).includes(feature))
			fail("INVALID_FRAME", "unknown protocol feature", `${path}[${i}]`);
	return result as ProtocolFeature[];
}
export function isCapability(value: unknown): value is DeviceCapability {
	return typeof value === "string" && (DEVICE_CAPABILITIES as readonly string[]).includes(value);
}
export function isProtocolFeature(value: unknown): value is ProtocolFeature {
	return typeof value === "string" && (PROTOCOL_FEATURES as readonly string[]).includes(value);
}

File: apps/ios/HostWire/Sources/HostWire/Capabilities.swift
import Foundation

/// Device capabilities a client may hold (capabilities.DEVICE_CAPABILITIES).
public enum DeviceCapability: String, Codable, CaseIterable, Sendable {
    case sessionsRead = "sessions.read"
    case sessionsPrompt = "sessions.prompt"
    case sessionsControl = "sessions.control"
    case sessionsManage = "sessions.manage"
    case bashRun = "bash.run"
    case terminalOpen = "term.open"
    case terminalInput = "term.input"
    case terminalResize = "term.resize"
    case filesRead = "files.read"
    case filesWrite = "files.write"
    case filesList = "files.list"
    case filesDiff = "files.diff"
    case agentsControl = "agents.control"
    case auditRead = "audit.read"
    case configRead = "config.read"
    case catalogRead = "catalog.read"
    case configWrite = "config.write"
    case brokerRead = "broker.read"
    case usageRead = "usage.read"
    case previewRead = "preview.read"
    case previewControl = "preview.control"
    case previewInput = "preview.input"
    case ciTrigger = "ci.trigger"
}

/// Negotiable protocol features (capabilities.PROTOCOL_FEATURES). The wire
/// carries these as strings; a newer client may name an unknown feature, so
/// requested-feature lists tolerate unknown values while granted lists are
/// strict (see `Capabilities.strictFeatures`).
public enum ProtocolFeature: String, Codable, CaseIterable, Sendable {
    case resume
    case hostWatch = "host.watch"
    case sessionWatch = "session.watch"
    case sessionState = "session.state"
    case sessionDelta = "session.delta"
    case sessionObserver = "session.observer"
    case sessionUnverified = "session.unverified"
    case sessionFork = "session.fork"
    case controllerLease = "controller.lease"
    case promptLease = "prompt.lease"
    case promptImages = "prompt.images"
    case transcriptImages = "transcript.images"
    case transcriptSearch = "transcript.search"
    case transcriptPage = "transcript.page"
    case projectReveal = "project.reveal"
    case agentLifecycle = "agent.lifecycle"
    case agentProgress = "agent.progress"
    case agentEvent = "agent.event"
    case agentTranscript = "agent.transcript"
    case terminalIO = "terminal.io"
    case filesList = "files.list"
    case filesSearch = "files.search"
    case filesDiff = "files.diff"
    case auditTail = "audit.tail"
    case catalogMetadata = "catalog.metadata"
    case settingsMetadata = "settings.metadata"
    case previewControl = "preview.control"
    case runtimeAdapters = "runtime.adapters"
    case workspaceLifecycle = "workspace.lifecycle"
    case clusterOperator = "cluster.operator"
}

/// `{ client: string[], server?: string[] }` — every entry must be a known
/// device capability (capabilities.decodeCapabilities).
public struct Capabilities: Codable, Equatable, Sendable {
    public let client: [String]
    public let server: [String]?

    public init(client: [String], server: [String]? = nil) {
        self.client = client
        self.server = server
    }

    private enum CodingKeys: String, CodingKey { case client, server }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let cl = try c.decode([String].self, forKey: .client)
        let sv = try c.decodeIfPresent([String].self, forKey: .server)
        try Capabilities.validate(cl, path: "client")
        if let sv { try Capabilities.validate(sv, path: "server") }
        client = cl
        server = sv
    }

    private static let known = Set(DeviceCapability.allCases.map(\.rawValue))

    /// Reject unknown capabilities (strict — used for capabilities + granted).
    public static func validate(_ values: [String], path: String) throws {
        for value in values where !known.contains(value) {
            throw T4WireError.invalidFrame(path: "capabilities.\(path)", reason: "unknown device capability \(value)")
        }
    }

    /// Additive feature list (capabilities.decodeFeatureList): bounded strings,
    /// unknown names tolerated so a newer client can negotiate forward.
    public static func requestedFeatures(_ values: [String], path: String) throws -> [String] {
        try values.enumerated().map { index, value in
            try Bounded.controlFree(value, path: "\(path)[\(index)]", maxBytes: 128)
        }
    }

    /// Strict feature list (capabilities.decodeNegotiatedFeatureList): every
    /// entry must be a known protocol feature.
    public static func strictFeatures(_ values: [String], path: String) throws -> [ProtocolFeature] {
        let knownFeatures = Set(ProtocolFeature.allCases.map(\.rawValue))
        return try values.enumerated().map { index, value in
            _ = try Bounded.controlFree(value, path: "\(path)[\(index)]", maxBytes: 128)
            guard knownFeatures.contains(value) else {
                throw T4WireError.invalidFrame(path: "\(path)[\(index)]", reason: "unknown protocol feature \(value)")
            }
            return ProtocolFeature(rawValue: value)!
        }
    }
}
